Mathlib/Algebra/InverseLimit.lean

```diff
@@ -0,0 +1,174 @@
+
+/-!
+# Inverse Limits of Modules, Abelian Groups, and Rings
+
+## Main definitions
+
+* `FooBar`
+
+## Main statements
+
+* `fooBar_unique`
+
+## Tags
+
+Foobars, barfoos
+-/
+
+
+variable {ι : Type*} [Preorder ι]
+variable (G : ι → Type*)
+
+
+section InverseSystem
+
+variable (f : ∀ i j, i ≤ j → G j → G i)
+
+/-- An inverse system is a family of objects and compatible morphisms,
+has a greater index than the codomain. -/
+class InverseSystem : Prop where
+  map_self' : ∀ i x, f i i le_rfl x = x
+  map_map' : ∀ {i j k} hij hjk x, f i j hij (f j k hjk x) = f i k (le_trans hij hjk) x
+
+variable {G} [InverseSystem G f]
+
+theorem InverseSystem.map_self i x : f i i le_rfl x = x :=
+  InverseSystem.map_self' i x
+
+theorem InverseSystem.map_map {i j k} (hij hjk x) :
+    f i j hij (f j k hjk x) = f i k (le_trans hij hjk) x :=
+  InverseSystem.map_map' hij hjk x
+
+end InverseSystem
+
+
+namespace Module
+
+variable (R : Type*) [Ring R]
+variable [∀ i, AddCommMonoid (G i)] [∀ i, Module R (G i)]
+variable (f : ∀ i j, i ≤ j → G j →ₗ[R] G i)
+
+private def inverseLimit : Submodule R (∀ i, G i) where
+  carrier := { a | ∀ i j hij, a i = f i j hij (a j) }
+  add_mem' ha hb i j hij := by simp [ha i j hij, hb i j hij]
+  zero_mem' := by simp
+  smul_mem' _ _ ha i j hij := by simp [ha i j hij]
+
+/-- The inverse limit of an inverse family of rings and ring homomorphisms. -/
+def InverseLimit : Type _ := inverseLimit G R f
+
+namespace InverseLimit
+
+variable {G R f}
+
+instance addCommMonoid : AddCommMonoid (InverseLimit G R f) :=
+  Submodule.addCommMonoid _
+
+instance module : Module R (InverseLimit G R f) :=
+  Submodule.module _
+
+@[ext]
+lemma ext {x y : inverseLimit G R f} (h : ∀ i, x.val i = y.val i) : x = y :=
+  Subtype.eq <| funext h
+
+end InverseLimit
+
+end Module
+
+
+namespace AddCommGroup
+
+variable [∀ i, AddCommGroup (G i)]
+variable (f : ∀ i j, i ≤ j → G j →+ G i)
+
+private def inverseLimit : AddSubgroup (∀ i, G i) :=
+  Submodule.toAddSubgroup <| Module.inverseLimit G ℤ
+    fun i j hij => AddMonoidHom.toIntLinearMap (f i j hij)
+
+def InverseLimit : Type _ := inverseLimit G f
+
+namespace InverseLimit
+
+variable {G f}
+
+instance addCommGroup : AddCommGroup (InverseLimit G f) :=
+  AddSubgroup.toAddCommGroup _
+
+@[ext]
+lemma ext {x y : inverseLimit G f} (h : ∀ i, x.val i = y.val i) : x = y :=
+  Subtype.eq <| funext h
+
+end InverseLimit
+
+end AddCommGroup
+
+
+namespace Ring
+
+variable [∀ i, Ring (G i)]
+variable (f : ∀ i j, i ≤ j → G j →+* G i)
+
+private def inverseLimit : Subring (∀ i, G i) where
+  carrier := (AddCommGroup.inverseLimit G fun i j hij => f i j hij).carrier
+  mul_mem' ha hb i j hij := by simp [ha i j hij, hb i j hij]
+  one_mem' := by simp [AddCommGroup.inverseLimit, Module.inverseLimit]
+  add_mem' := (AddCommGroup.inverseLimit G fun i j hij => f i j hij).add_mem'
+  zero_mem' := (AddCommGroup.inverseLimit G fun i j hij => f i j hij).zero_mem'
+  neg_mem' ha i j hij := by simp [ha i j hij]
```

Comment:
You can do something like
```suggestion
  __ := AddCommGroup.inverseLimit (G · · ·)
  mul_mem' ha hb i j hij := by simp [ha i j hij, hb i j hij]
  one_mem' := by simp [AddCommGroup.inverseLimit, Module.inverseLimit]
```

Reply:
Thanks! I knew something like this was possible but couldn't remember the syntax.